src/lib/protocols/CHIPServerBase.cpp

```diff
@@ -0,0 +1,239 @@
+ *    Copyright (c) 2013-2017 Nest Labs, Inc.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file implements a common, non-instantiatable base object
+ *      for implementing CHIP protocol unsolicited responders
+ *      (i.e. servers) that encapsulates validating authenticated
+ *      requests and sending status reports and also provides common
+ *      data member storage for fabric state and an exchange manager.
+ *
+ */
+
+#include "CHIPServerBase.h"
+#include "CHIPProtocols.h"
+
+#include <core/CHIPCore.h>
+#include <core/CHIPEncoding.h>
+#include <core/CHIPTLV.h>
+#include <protocols/common/CommonProtocol.h>
+#include <support/CodeUtils.h>
+
+namespace chip {
+namespace Protocols {
+
+using namespace chip::Encoding;
+using namespace chip::TLV;
+
+/**
+ * Determine whether an incoming request message to a CHIP server should be accepted or discarded.
+ *
+ * This method is intended to be used by CHIP server implementations to implement extensible access control
+ * policy for incoming request messages.  Server implementations that rely on delegate objects should call
+ * this method early in message processing to determine whether message processing should continue.
+ *
+ * This method calls the virtual ChipServerDelegateBase::EnforceAccessControl() method on the supplied delegate to evaluate access
+ * control policy for the message.  CHIP server delegate classes, and application-specific delegates derived
+ * from the standard server classes, should override the virtual method to enforce specific access control
+ * policies.
+ *
+ * @param[in] ec            The ExchangeContext over which the message was received.
+ * @param[in] msgProfileId  The profile id of the received message.
+ * @param[in] msgType       The message type of the received message.
+ * @param[in] msgInfo       A ChipMessageInfo structure containing information about the received message.
+ * @param[in] delegate      The delegate object supplied by the application that can be used to override
+ *                          the default message access control policy.
+ *
+ * @retval true             If the message should be accepted and processed as normal.
+ * @retval false            If message processing should stop and the message the should be discarded.
+ */
+bool ChipServerBase::EnforceAccessControl(ExchangeContext * ec, uint32_t msgProfileId, uint8_t msgType,
```

Comment:
would prefer that these classes did not begin with 'Chip', as they're already namespaced into chip::